toolchain/check/eval_inst.cpp

```diff
@@ -264,6 +264,57 @@ auto EvalConstantInst(Context& context, SemIR::InstId inst_id,
     // TODO: Add note pointing to the impl declaration.
     context.emitter().Emit(inst_id, ImplAccessMemberBeforeSet);
     return ConstantEvalResult::Error;
+  } else if (auto witness = context.insts().TryGetAs<SemIR::LookupImplWitness>(
+                 inst.witness_id)) {
+    // If the witness is symbolic but has a self type that is a FacetType, it
+    // can pull rewrite values from the self type. If the access is for one of
+    // those rewrites, evaluate to the RHS of the rewrite.
+
+    auto witness_self_type_id =
+        context.insts().Get(witness->query_self_inst_id).type_id();
+    if (!context.types().Is<SemIR::FacetType>(witness_self_type_id)) {
+      return ConstantEvalResult::NewSamePhase(inst);
+    }
+
+    // The `ImplWitnessAccess` is accessing a value, by index, for this
+    // interface.
+    auto access_interface_id = witness->query_specific_interface_id;
+
+    auto witness_self_facet_type_id =
+        context.types()
+            .GetAs<SemIR::FacetType>(witness_self_type_id)
+            .facet_type_id;
+    // TODO: We could consider something better than linear search here, such as
+    // a map. However that would probably require heap allocations which may be
+    // worse overall since the number of rewrite constraints is generally low.
+    for (auto witness_rewrite : context.facet_types()
+                                    .Get(witness_self_facet_type_id)
+                                    .rewrite_constraints) {
+      // Look at each rewrite constraint in the self facet value's type. If the
+      // LHS is an `ImplWitnessAccess` into the same interface that `inst` is
+      // indexing into, then we can use its RHS as the value.
+      if (auto witness_rewrite_lhs_access =
+              context.insts().TryGetAs<SemIR::ImplWitnessAccess>(
+                  witness_rewrite.lhs_id)) {
+        if (witness_rewrite_lhs_access->index != inst.index) {
+          continue;
+        }
```

Comment:
Maybe restructure to avoid nesting, e.g.:

```suggestion
      auto witness_rewrite_lhs_access =
          context.insts().TryGetAs<SemIR::ImplWitnessAccess>(
              witness_rewrite.lhs_id);
      if (!witness_rewrite_lhs_access || witness_rewrite_lhs_access->index != inst.index) {
        continue;
      }
```

Note this'd imply removing one of the braces below...

Reply:
Done